src/pathfinder/water_regions.cpp

```diff
@@ -197,25 +197,29 @@ class WaterRegion
 
 		const size_t max_element_width = std::to_string(this->data.number_of_patches).size();
 
-		std::array<int, 16> traversability_NW{0};
-		for (auto bitIndex : SetBitIterator(this->data.edge_traversability_bits[DIAGDIR_NW])) *(traversability_NW.rbegin() + bitIndex) = 1;
-		Debug(map, 9, "    {:{}}", fmt::join(traversability_NW, " "), max_element_width);
+		std::string traversability = fmt::format("{:0{}b}", this->data.edge_traversability_bits[DIAGDIR_NW], WATER_REGION_EDGE_LENGTH);
+		Debug(map, 9, "    {:{}}", fmt::join(traversability, " "), max_element_width);
 		Debug(map, 9, "  +{:->{}}+", "", WATER_REGION_EDGE_LENGTH * (max_element_width + 1) + 1);
 
+		/* The region's patch labels may be empty, in which case a dummy array is used. */
+		static const TWaterRegionPatchLabelArray empty_labels{};
+		static_assert(INVALID_WATER_REGION_PATCH == TWaterRegionPatchLabel{});
+
+		const auto &labels = this->data.tile_patch_labels == nullptr ? empty_labels : *this->data.tile_patch_labels;
+		auto it = std::begin(labels);
 		for (int y = 0; y < WATER_REGION_EDGE_LENGTH; ++y) {
 			std::string line{};
 			for (int x = 0; x < WATER_REGION_EDGE_LENGTH; ++x) {
-				const auto label = this->GetLabel(TileAddXY(tile_area.tile, x, y));
+				const auto label = *it++;
```

Comment:
While slightly more efficient, I do feel that this makes the code harder to understand. In addition the new code relies on more knowledge implementation details, such as the order of tile iteration and the fact that we only conditionally store tile label data. If any of that changes, this code will break, either by not compiling or silently.

Since this is just debug printing code, I wouldn't worry too much about a little performance waste.